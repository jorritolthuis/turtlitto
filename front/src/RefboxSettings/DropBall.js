import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";

/**
 * Constructs a button for the Drop Ball command
 * Author: G.W. van der Heijden
 * Author: T.T.P. Franken
 *
 * Props:
 * - onClick: a function on what to do when the button is pressed
 */
const DropBall = props => {
  return (
    <DropBallWrapper className={props.className}>
      <DropBallButton
        id={"drop-ball-button"}
        onClick={() => {
          props.onClick();
        }}
      >
        DB
      </DropBallButton>
    </DropBallWrapper>
  );
};

const DropBallWrapper = styled.div`
  height: 25%;
  display: flex;
  flex-direction: row;
  justify-content: center;
`;

const DropBallButton = styled.button`
  background-color: ${props => props.theme.button};
  height: 75%;
  width: 25%;
  border: 0.25rem solid;
  font-size: 2rem;
  margin: auto;
  &:active {
    background-color: ${props => props.theme.buttonActive};
  }
`;

DropBall.propType = {
  onClick: PropTypes.func.isRequired
};

export default DropBall;
