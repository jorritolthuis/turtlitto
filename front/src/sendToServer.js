export default (message, destination, session) => {
  const l = window.location;
  const msg = JSON.stringify(message);
  console.log(`send ${msg} to ${l.protocol}//${l.host}/api/v1/${destination}`);
  return fetch(`${l.protocol}//${l.host}/api/v1/${destination}`, {
    method: "POST",
    headers: new Headers({
      Authorization: "Basic " + btoa(`user:${session}`)
    }),
    body: msg
  });
};
