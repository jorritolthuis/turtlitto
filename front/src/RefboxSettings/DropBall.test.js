import React from "react";
import DropBall from "./DropBall";
import { shallow } from "enzyme";
import sendToServer from "../sendToServer";

describe("DropBall", () => {
  it("shows a button with DB written on it", () => {
    const wrapper = shallow(<DropBall onClick={() => {}} />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe("When clicked, the Drop Ball button", () => {
  const realFetch = global.fetch;
  let wrapper = null;
  beforeEach(() => {
    wrapper = shallow(
      <DropBall
        onClick={() => {
          sendToServer("dropped_ball", "command");
        }}
      />
    );
    const l = window.location;
    global.fetch = jest.fn().mockImplementation((url, params) => {
      expect(url).toBe(`${l.protocol}//${l.host}/api/v1/command`);
      expect(params).toMatchSnapshot();
      return Promise.resolve({ ok: true });
    });
  });

  afterEach(() => {
    global.fetch = realFetch;
  });
  it("should pass `dropped_ball` to the 'sendToServer' function", () => {
    wrapper.find("#drop-ball-button").simulate("click");
  });
});
