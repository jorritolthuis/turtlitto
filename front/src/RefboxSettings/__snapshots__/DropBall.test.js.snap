// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropBall shows a button with DB written on it 1`] = `
<styled.div>
  <styled.button
    id="drop-ball-button"
    onClick={[Function]}
  >
    DB
  </styled.button>
</styled.div>
`;

exports[`When clicked, the Drop Ball button should pass \`dropped_ball\` to the 'sendToServer' function 1`] = `
Object {
  "body": "\\"dropped_ball\\"",
  "headers": Object {},
  "method": "PUT",
}
`;
